### DISCUSSION TITLE
Contributing?

### DISCUSSION THREAD

**Discussion by mixonomicon** [morelinq/MoreLINQ#1072] [General]
Hi,

I made some extensions of my own, which I think might benefit the community. Since you guys already have this thing going, I figure its better if I contribute than to make my own package. 

Now, how does one contribute to this project?

I forked, and setup the devenv, however, for start you have `TreatWarningsAsErrors` set and there seem to be a dozen code-styling warnings that are now preventing the build (IDE0036, IDE0058, etc), making me wonder how you build the code while working on it. csprojs seem to be very convoluted at this point, most of it seems part of some legacy build processes. But its fine, not looking to rain on that, I just want to know how to setup the work environment locally so I can add extensions and tests and create a PR for you. There does not seem to be any CONTRIBUTING document explaining things.  

Any help is appreciated.

**Comment by viceroypenguin** [comment]
Hi @the-black-wolf - I maintain a fork that I strive to be more accessible and active. You're welcome to submit some ideas to https://github.com/viceroypenguin/SuperLinq/, and I'd be glad to review and hopefully include in SuperLinq. I hope you are able to contribute your ideas to MoreLinq at some point, but this would be a good alternative in the meantime.

**Comment by atifaziz** [comment]
> I made some extensions of my own, which I think might benefit the community. Since you guys already have this thing going, I figure its better if I contribute than to make my own package.

That's a great call and thanks for reaching out.
 
> I forked, and setup the devenv, however, for start you have `TreatWarningsAsErrors` set and there seem to be a dozen code-styling warnings that are now preventing the build (IDE0036, IDE0058, etc), making me wonder how you build the code while working on it.

There's a [`global.json`](https://github.com/morelinq/MoreLINQ/blob/9c130ff557657f26132c6f03961b0454029fecf7/global.json) that says which version of the SDK you should use (latest patches are okay). If you have that, there should be no build errors. You can build on the command line with [`build.cmd`](https://github.com/morelinq/MoreLINQ/blob/9c130ff557657f26132c6f03961b0454029fecf7/build.cmd) on Windows or [`build.sh`](https://github.com/morelinq/MoreLINQ/blob/9c130ff557657f26132c6f03961b0454029fecf7/build.sh) for most other platforms. Alternatively, just open [the solution file](https://github.com/morelinq/MoreLINQ/blob/9c130ff557657f26132c6f03961b0454029fecf7/MoreLinq.sln) in Visual Studio and build.

If you delete [`global.json`] and have a later SDK version installed then there may be new analyzers or regressions that will flag errors. Those will need to be filed and addressed.

> csprojs seem to be very convoluted at this point, most of it seems part of some legacy build processes.

You shouldn't have to go in there most of the time, but I'd like to understand what parts appear specifically convoluted.

> There does not seem to be any CONTRIBUTING document explaining things.

Yes, that's sorely missing, but contributing requires a pretty common sense approach:

1. Make sure you're not duplicating [something that's already there](https://github.com/morelinq/MoreLINQ/blob/9c130ff557657f26132c6f03961b0454029fecf7/README.md#operators).
2. Have a quick browse through the [issues](https://github.com/morelinq/MoreLINQ/issues?q=is%3Aissue) to see what you have in mind isn't already proposed.
3. If you fell you have a new idea and unique idea to contribute, [open an issue](https://github.com/morelinq/MoreLINQ/issues/new) and describe your proposal (ideally with examples). Often an idea will require a discussion, detailing, challenging and fine-tuning before it can be submitted as a PR for review.

Here are some guidelines that will help up the chances for ideas getting accepted:

1. Contribute something that's algorithmically rich, generic and broadly applicable.
1. Avoid simple helpers that merely act as wrappers of other operations for cosmetics.
1. Follow the style and conventions of the project. They're there for a good reason.
1. Unless you're in for the long haul, keep in mind that maintainers will be committing to your code and maintaining it for potentially many years to come (that's already been the case given the project's long history). Consequently, don't be disheartened if some of your ideas aren't accepted. There can be a multitude of reasons, most of which will be discussed as part of a proposal.
1. Have patience. People have busy schedules. Time is scarce and so things can take time to cook. This can be frustrating, but this is open source and no one is pointing a gun to anyone's head.
1. Contribute because you want to be part of something bigger.

**Comment by atifaziz** [comment]
> a dozen code-styling warnings that are now preventing the build (IDE0036, IDE0058, etc)

@the-black-wolf This is being tracked in #1073 and hopefully addressed by PR #1074 so you should be able to build on the latest version of the .NET SDK without any issues.

**Reply by mixonomicon** [reply]
I've temporarily disabled the TreatWarningAsError so I can build.
